### ISSUE TITLE
[Variables and calculated Waypoints] enhance featureset of Formula and/or Coordinate Parser

### ISSUE THREAD

**Issue by eddiemuc** [cgeo/cgeo#12316]
### Tell us your idea!

Global variables use a formula syntax to parse and interpret expressions containing variables. This syntax is documented here: https://github.com/cgeo/cgeo/wiki/Calculator-Formula-Syntax 
Calculated coordinates use an own parser build on top of this to parse coordinates. This syntax is not documented yet (ToDo), but generally speaking the algorithm tries to find separate valid "expressions" (as described above) separated by whitespace, °, ' or " and interpret the result as a coordinate (special handling for N, E, O, S, W integrated).

In dev meeting as of 12.12.21 is was suggested to test and (where needed) further enhance the parsing syntax for both variable formulas and/or coordinate parser. This issue is to collect and discuss further possible improvements.

* Formula expression syntax can be tested by opening any cache in nightly or dev build, go to variable page and start typing. Tap "help" for syntax explanation (will direct you to above linked wiki page)
* Coordinate parser expression syntax can be tested by creating any waypoint, opening "calculated coordinates (global)" and start typing

### What solution would you suggest?

The following suggestions have already been made (@ztNFny , @murggel : I tried to capture your suggestions from dev meeting, please correct if wrong):

### Ignore space in concat expressions
For example the expression `12 .(A+1) A 4` (with spaces) should parse correctly as if it were `12.(A+1)A4`. 

This will work for variable formulas, but would be more problematic in coordinate parsing. Problem is that in coordinate parsing space then can't be used as separator between coordinate parts any more. For example, take coordinate expression `N4B 12 .(A+1) A 4`. How should parser know that the first space (after N4B) separates degree from minutes, but the other spaces should be concatennated into one minute expression?

### Allow certain characters as additional indicators for operations
For example, interpret `:` as division in addition to `/`, interpret `x` as multiplicator in addition to `*`, interpret `—` (dash) as minus.

TODO: I would need suggestions for that. 

Some chars are easy to add, but things like `x` are problematic since they might be interpreted as a variable as well, especially in combination with the first request, I don't think x will work). For example. consider expression `4 x 5`. Should this mean `4 * 5` (evaluates to `20`) or should this mean `4x5` (with `x=1` evaluates to `415`). Anyone has an idea?

**Comment by ztNFny** [comment]
For minus: – – —
For muplication: •
Brackets: ([ ])

As for "x": It's not possible to always get it right, but I think it's more common for it to be used as multiplication char than as variable name (those "usually" start at A, so getting to X is a long way). Maybe interpret a "x" as multiplication and "X" as variable? If it's wrong the user can still change it (and will have to check it anyway as formula recognition will always be hit and miss)

**Comment by eddiemuc** [comment]
Thanks @ztNFny , I will add the "easy-addeable" symbols to `Formula` class.

Give me some time to chew on the "x" thing.... Generally forbidding the "x" (lowercase) as a usable variable name (or start of a usable variable name) might be a way to go, let me think about possible side effects. E.g. something like the following would then result in a parser error: `5 * x` (`5 * X` would work though)

**Comment by Lineflyer** [comment]
Regarding the `x` I also think there are only two possibilities:
- Leave out `x` as variable and make it usable as multiplication mark
- Stick with `*` for multiplications

I would avoid different meanings between lowercase and uppercase...it not obvious
I would suggest to stick with `*` for multiplications mainly because x is indeed used in many owner provided formulas as a variable. Furthermore `*` is IMHO well known as multiplication sign (and already now used by the existing calculator).

**Comment by ztNFny** [comment]
> I would suggest to stick with * for multiplications

It's not about what we use in our internal formulas but about recognizing formulas from a listing. So if a listing contains "A = B x C" should that be recognized as formula: "A=BxC" or "A=B*C"

**Comment by ztNFny** [comment]
> Generally forbidding the "x" (lowercase) as a usable variable name (or start of a usable variable name) might be a way to go, let me think about possible side effects. 

That's not what I meant to suggest. I was only meaning it in the content of recognizing the formula from listing. We should NOT forbid using a certain char or interpret a MANUALLY entered x as *.

**Comment by eddiemuc** [comment]
> > Generally forbidding the "x" (lowercase) as a usable variable name (or start of a usable variable name) might be a way to go, let me think about possible side effects.
> 
> That's not what I meant to suggest. I was only meaning it in the content of recognizing the formula from listing. We should NOT forbid using a certain char or interpret a MANUALLY entered x as *.

Thanks for clearing that up, this was not clear to me as well.

Currently there is no distinction in the formula input field wrt where a formula comes from (e.g. whether the user entered it by typing in the keyboard or whether he copied it from a listing). There is no special logic to recognize other formulas and convert them to a c:geo usable format. 

If the syntax for formulas interpretable by c:geo and the formulas recognized as such from listings start to differ (and, to take the next step, there should be some sort of conversion from "external formulas" to "internal formulas", then soon two separate parsers need to be available for these use cases. We really want to introduce this kind of complexity?

**Comment by ztNFny** [comment]
> We really want to introduce this kind of complexity?

From my experience with writing formula parsers I think it's actually less complex to do it this way. If you have one conversion steps with like (x|*|• -> *) conversion then in all your calculation logic you can just throw those strings into a normal math parser without worrying about special cases. So you have one complex step (listing parsing) - which you have to have anyway - and one simple calculation step. Otherwise you need to have the complex parsing everywhere in your calculation.

For the manual formula input it's then as easy as saying +-*/... are calculation char, everything else is a variable. So a manually entered x is a variable, just like an a.
If the input fields don't follow some very strict logic it'll be very hard to properly input "complex" formulas as it's not WYSIWYG.

**Comment by ztNFny** [comment]
To clarify, I suggest to have a listing parser like:
formulaFromListing("A ÷ B x C") { return "A/B*C" }
and then insert that result into the waypoint fields. Maybe there also needs to be a button "parse this text as formula" for text not recognized from listing but copy & pasted manually.
and then in the "actual" calculation only parse the "normal" calculation chars

btw, another char from a real-world listing: "÷" for "/"

**Comment by eddiemuc** [comment]
> To clarify, I suggest to have a listing parser like: formulaFromListing("A ÷ B x C") { return "A/B*C" } and then insert that result into the waypoint fields. 

That's understood. But there is no such parser in c:geo yet. (and up to now there is no such parser planned).

What is available is this:
* A `Formula` class (formerly known as `Calculator`) which parses and evaluates Strings as expressions returning single, "primitive" values (number or text) 
* A `DegreeFormula` class which parses and evaluates Strings supposed to contain a degree part of a coordinate. It uses Formula under the hood
* A "listing scanner" method to scan listings for possible formula - but this "scanner" is a very simple regular-expression parser, not supposed to evaluate the deeper insides of a formula (e.g. results are not necessarily parseable formulas)

It looks to me like you want to highly "upgrade" the "listing scanner" method, is that correct? My fear is that this method over time will become close to Formula in complexity and we have to maintain two similar complex parser/scanner logics in parallel.

> btw, another char from a real-world listing: "÷" for "/"

Ok, easy to add (to Formula)

**Comment by ztNFny** [comment]
> My fear is that this method over time will become close to Formula in complexity and we have to maintain two similar complex parser/scanner logics in parallel.

I only want 1 complex parser method, the formulaFromListing() one.
Then Formula can stay quite primitive as it only has to evaluate strings in a known format. So we could just throw a string to exp4j or whatever formula parser we want to use (I didn't look at the existing code, so no idea how it's done right now) and not care about anything at this stage.

If you do the "intelligent" parsing of the formula in Formula then the user has no control over what happens there (whether x is x or *) - and as there's so many different way to write formulas imho all the parsing must be verifyable and correctable by the user, so it has to be an in-between step.

We talked about GCBuddy yesterday - isn't it the same there? In the formula dialog you have their own keyboard which limits the chars that you can input so you can only produce formulas in a known format that are easy to parse for the app.

**Comment by eddiemuc** [comment]
@ztNFny I think I understand now.

`Formula` is self-written right now from scratch. No lib like exp4j is used in the background. The reason for that is that formulas used in conjunction with geocaching have some different semantics than in the rest of the world. The best example for this is something I call "concatenation":

Consider following expression:  `21A5`
Assuming A is a variable with value 2, then most libs I know (and also most mathematicians I know) would evaluate such an expression as 21*2*5 which evaluates to `210`. However, in geocaching context such an expression is usually evaluated in a concattenated form, which means it evaluates to `2125`.

What is complex and what not is always subjective, but I would call Formula pretty complex, and at the same time pretty powerful right now. But it is somewhat specific for geocaching needs.

So right now it is implemented the other way around: the complexity and "intelligence" is in the  Formula parser who parses the user input, and the "scanning" is very primitive. So far, I was able to keep the Formula syntax both flexible and unambigious. There might come a point where this is no longer possible though -> then some things have to be moved to the "scanner" perhaps.

Maybe the "x" discussion is the first such case. It could be solved in the "scanner" method very easy and primitive: replace occurences of " x " (with space before and after) with " * " before giving it to Formula. Maybe this is the right way to solve this. 

I looked at GCBuddy before implementing all this, I haven't found a distinction between a "scan parser" and a "formula parser" there. Especially the whole concattenation logic is implemented in the "normal" user input parser there.

Then there are some very c:geo specific extensions (c:geo likes to make everything a bit more complex than other apps :-)):
* The overflow character _ which is needed solely to mimic the strange "calculated coordinate" inputs from the single-char-masks (something GCBuddy does not have - only "Plain input" is supported)
* The whole add/remove variable logic (something GCBuddy does not have - there are vars from A to Z and that's that)
* The support for var names with > 1 char (something GCBuddy does not have - there are vars from A to Z and that's that)

**Comment by eddiemuc** [comment]
> Maybe off-topic: https://github.com/siggel/coordinatejoker offers a nice possibility to test multiple input values and see the resulting waypoints. Would be nice to have this (needs to be a separate Feature request), I'm bringing it up now as it might need some calculation stuff in a specific place and gives additional input in that regard.

Yeah, would love this directly in c:geo :-) Next thing after the whole calc coordinate stuff is in place...

**Comment by eddiemuc** [comment]
The more easy improvements are part of PR #12318

**Comment by eddiemuc** [comment]
The harder thing is now part of PR #12368
